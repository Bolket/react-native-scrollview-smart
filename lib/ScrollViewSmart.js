import React, {
  Component,
} from 'react';
import PropTypes from 'prop-types';
import {
  Platform,
  Keyboard,
  ScrollView,
  View,
} from 'react-native';

const propTypes = {
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
  extraSpace: PropTypes.number,
  nodeHeight: PropTypes.number,
  tabFooter: PropTypes.number,
};

const defaultProps = {
  extraSpace: 0,
  nodeHeight: 64,
  tabFooter: 0,
};

class ScrollViewSmart extends Component {

  constructor(props) {
    super(props);

    this.state = {
      keyboardSpace: 0,
    };
    this.inputFocused = this.inputFocused.bind(this);
    this.updateKeyboardSpace = this.updateKeyboardSpace.bind(this);
    this.resetKeyboardSpace = this.resetKeyboardSpace.bind(this);
  }

  componentDidMount() {
    if (Platform.OS === 'android') {
      this.keyboardWillShowEvent = Keyboard.addListener(
        'keyboardDidShow',
        this.updateKeyboardSpace,
      );
      this.keyboardWillHideEvent = Keyboard.addListener(
        'keyboardDidHide',
        this.resetKeyboardSpace,
      );
    } else {
      this.keyboardWillShowEvent = Keyboard.addListener(
        'keyboardWillShow',
        this.updateKeyboardSpace,
      );
      this.keyboardWillHideEvent = Keyboard.addListener(
        'keyboardWillHide',
        this.resetKeyboardSpace,
      );
    }
  }

  componentWillUnmount() {
    this.keyboardWillShowEvent.remove();
    this.keyboardWillHideEvent.remove();
  }

  inputFocused(reactNode) {
    if (Platform.OS === 'ios') {
      setTimeout(() => {
        const scrollResponder = this.refs.scrollView.getScrollResponder();
        scrollResponder.scrollResponderScrollNativeHandleToKeyboard(
          reactNode,
          this.props.nodeHeight,
          true,
        );
      }, 50);
    }
  }

  updateKeyboardSpace(frames) {
    if (frames.end) {
      this.setState({ keyboardSpace: frames.end.height + this.props.extraSpace });
    } else {
      this.setState({ keyboardSpace: frames.endCoordinates.height + this.props.extraSpace });
    }
  }

  resetKeyboardSpace() {
    this.setState({
      keyboardSpace: 0,
    });
  }

  render() {
    return (
      <ScrollView
        ref={'scrollView'}
        keyboardShouldPersistTaps="always"
        keyboardDismissMode={'interactive'}
        showsVerticalScrollIndicator={false}
        {...this.props}
      >
        {this.props.children}
        <View style={{ height: this.state.keyboardSpace + this.props.tabFooter }} />
      </ScrollView>
    );
  }
}

ScrollViewSmart.propTypes = propTypes;
ScrollViewSmart.defaultProps = defaultProps;

export default ScrollViewSmart;
