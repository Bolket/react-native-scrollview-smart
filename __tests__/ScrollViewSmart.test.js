/**
 * eslint-disable
 */

import { View, TextInput } from 'react-native';
import React from 'react';
import { expect as expectChai, should } from 'chai';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ScrollViewSmart from '../lib/ScrollViewSmart';

describe('ScrollViewSmart', () => {
  let component;
  let bottomView;
  let textInput;

  it('renders correctly', () => {
    const tree = renderer.create(<ScrollViewSmart />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  beforeEach(() => {
    component = shallow(
      <ScrollViewSmart>
        <View style={{ flex: 1, marginTop: 50 }}>
          <TextInput />
        </View>
      </ScrollViewSmart>
    );
    bottomView = component.find('View');
    textInput = component.find('TextInput');
  });

  it('it should render 2 View component', () => {
    expectChai(bottomView).to.have.length(2);
  });

  it('it should render 1 TextInput component', () => {
    expectChai(textInput).to.have.length(1);
  });

  it('should have props for title and address', () => {
    expectChai(component.props().extraSpace).to.equal(0);
    expectChai(component.props().nodeHeight).to.equal(64);
    expectChai(component.props().tabFooter).to.equal(0);
  });
});
