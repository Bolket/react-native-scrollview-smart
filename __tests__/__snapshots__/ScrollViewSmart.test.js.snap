// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollViewSmart renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <ScrollViewMock
    extraSpace={0}
    keyboardDismissMode="interactive"
    keyboardShouldPersistTaps="always"
    nodeHeight={64}
    showsVerticalScrollIndicator={false}
    tabFooter={0}
>
    <View
        style={
            Object {
                "flex": 1,
                "marginTop": 50,
              }
        }
    >
        <TextInput />
    </View>
    <View
        style={
            Object {
                "height": 0,
              }
        }
    />
</ScrollViewMock>,
  "nodes": Array [
    <ScrollViewMock
      extraSpace={0}
      keyboardDismissMode="interactive"
      keyboardShouldPersistTaps="always"
      nodeHeight={64}
      showsVerticalScrollIndicator={false}
      tabFooter={0}
>
      <View
            style={
                  Object {
                        "flex": 1,
                        "marginTop": 50,
                      }
            }
      >
            <TextInput />
      </View>
      <View
            style={
                  Object {
                        "height": 0,
                      }
            }
      />
</ScrollViewMock>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_element": <ScrollViewSmart
      extraSpace={0}
      nodeHeight={64}
      tabFooter={0}
>
      <View
            style={
                  Object {
                        "flex": 1,
                        "marginTop": 50,
                      }
            }
      >
            <TextInput />
      </View>
</ScrollViewSmart>,
    "_instance": ScrollViewSmart {
      "context": Object {},
      "inputFocused": [Function],
      "props": Object {
        "children": <View
          style={
                    Object {
                              "flex": 1,
                              "marginTop": 50,
                            }
          }
>
          <TextInput />
</View>,
        "extraSpace": 0,
        "nodeHeight": 64,
        "tabFooter": 0,
      },
      "refs": Object {},
      "resetKeyboardSpace": [Function],
      "state": Object {
        "keyboardSpace": 0,
      },
      "updateKeyboardSpace": [Function],
      "updater": Updater {
        "_renderer": [Circular],
      },
    },
    "_newState": null,
    "_rendered": <ScrollViewMock
      extraSpace={0}
      keyboardDismissMode="interactive"
      keyboardShouldPersistTaps="always"
      nodeHeight={64}
      showsVerticalScrollIndicator={false}
      tabFooter={0}
>
      <View
            style={
                  Object {
                        "flex": 1,
                        "marginTop": 50,
                      }
            }
      >
            <TextInput />
      </View>
      <View
            style={
                  Object {
                        "height": 0,
                      }
            }
      />
</ScrollViewMock>,
    "_rendering": false,
    "_updater": Updater {
      "_renderer": [Circular],
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ScrollViewSmart
    extraSpace={0}
    nodeHeight={64}
    tabFooter={0}
>
    <View
        style={
            Object {
                "flex": 1,
                "marginTop": 50,
              }
        }
    >
        <TextInput />
    </View>
</ScrollViewSmart>,
}
`;
